src/train.py

```diff
@@ -16,67 +16,248 @@
 import torch.nn as nn
 import torch
 import numpy as np
+from collections import defaultdict
 
 DEVICE = init_device()
```

Comment:
Not an issue for now, but maybe we can make `device` one of the `args` later one.

Reply:
Consider it done! I think instead of passing around the device variable everywhere, it is easier to just make it global in the main method.  We could also move it to the top of train if you think it is better suited there!